In [1]:
import pandas as pd
DATA_DIR = "../data/"
TRAINDATA_DIR = "../traindata/"

In [3]:
#read tweets

name_tweets = DATA_DIR + "clean_tweets/clean_tweets.json"
name_tweets_before = DATA_DIR + "clean_tweets/clean_tweets_before.json"
name_tweets_after = DATA_DIR + "clean_tweets/clean_tweets_after.json"

with open(name_tweets, "r") as file:
    tweets_df = pd.read_json(file, orient='index')
    
with open(name_tweets_before, "r") as file:
    before_tweets_df = pd.read_json(file, orient='index')
    
with open(name_tweets_after, "r") as file:
    after_tweets_df = pd.read_json(file, orient='index')
    
print(tweets_df.describe())

         text            timestamp                      tweet
count   13132                13132                      13132
unique  12729                13064                      13063
top     https  2018-07-15 13:33:15  MAKE AMERICA GREAT AGAIN!
freq       42                    5                         18
first     NaN  2015-06-14 21:15:20                        NaN
last      NaN  2018-09-26 10:57:30                        NaN


In [4]:
after_tweets_df

text           timestamp  \
0     thank you pennsylvania going to new hampshire ... 2016-11-08 00:08:28   
1     today in florida i pledged to stand with the p... 2016-11-08 00:16:15   
2     big news to share in new hampshire tonight pol... 2016-11-08 00:17:57   
3     unbelievable evening in new hampshire thank yo... 2016-11-08 03:43:54   
4     rt https https in michigan this is your moveme... 2016-11-08 04:27:18   
5     rt https thanks new hampshire nh newhampshire ... 2016-11-08 04:29:10   
6     today we are going to win the great state of m... 2016-11-08 06:42:36   
7                     today we make america great again 2016-11-08 11:43:14   
8     vote today go to https to find your polling lo... 2016-11-08 16:39:36   
9     we need your vote go to the polls let s contin... 2016-11-08 18:03:49   
10                              electionday https https 2016-11-08 18:23:39   
11    i will be watching the election results from t... 2016-11-08 21:18:04   
12    just out according to https utah officials rep... 2016-11-08 21:28:24   
13    don t let up keep getting out to vote this ele... 2016-11-08 21:31:20   
14    still time to votetrump ivoted electionnight h... 2016-11-08 23:03:42   
15    rt https final push eric and i doing dozens of... 2016-11-08 23:20:09   
16    rt https join my family in this incredible mov... 2016-11-08 23:20:39   
17    rt https such a surreal moment to vote for my ... 2016-11-09 01:35:15   
18    watching the returns at pm electionnight maga ... 2016-11-09 02:48:27   
19    such a beautiful and important evening the for... 2016-11-09 11:36:58   
20    happy st birthday to the u s marine corps than... 2016-11-10 19:31:27   
21    a fantastic day in d c met with president obam... 2016-11-11 02:10:46   
22    just had a very open and successful presidenti... 2016-11-11 02:19:44   
23    love the fact that the small groups of protest... 2016-11-11 11:14:20   
24    busy day planned in new york will soon be maki... 2016-11-11 13:33:35   
25    today we express our deepest gratitude to all ... 2016-11-11 15:26:37   
26    this will prove to be a great time in the live... 2016-11-12 15:05:33   
27    wow the https is losing thousands of subscribe... 2016-11-13 14:16:46   
28    the https sent a letter to their subscribers a... 2016-11-13 14:43:09   
29    mitt romney called to congratulate me on the w... 2016-11-13 14:45:28   
...                                                 ...                 ...   
5309  tiger is playing great looks like a big win co... 2018-09-23 20:43:52   
5310  going to new york will be with prime minister ... 2018-09-23 20:52:37   
5311  prime minster https is coming up to trump towe... 2018-09-23 22:48:33   
5312  rt https now is the time to put safety first a... 2018-09-24 14:13:22   
5313  remarks by president trump at global call to a... 2018-09-24 14:17:26   
5314  today we commit to fighting the drug epidemic ... 2018-09-24 16:55:18   
5315  it was my great honor to welcome and meet with... 2018-09-24 20:42:22   
5316  us korea free trade agreement signing ceremony... 2018-09-24 20:44:24   
5317  brett kavanaugh and his wife ashley will be in... 2018-09-24 21:33:33   
5318  joint statement on the united states korea fre... 2018-09-24 21:46:22   
5319  it was my great honor to welcome president htt... 2018-09-24 22:33:39   
5320  it was my great honor to welcome president htt... 2018-09-24 23:41:37   
5321                                              https 2018-09-25 00:22:46   
5322                              remember the midterms 2018-09-25 02:38:39   
5323  the democrats are working hard to destroy a wo... 2018-09-25 02:50:09   
5324  republican party favorability is the highest i... 2018-09-25 10:41:20   
5325  despite requests i have no plans to meet irani... 2018-09-25 10:53:41   
5326  will be speaking at the united nations this mo... 2018-09-25 13:14:04   
5327                                              https 2018-09-25 15:27:35   
5328                  

In [5]:
#read stock data

name_stock = DATA_DIR + "DJI-20years-daily-change.json"

with open(name_stock, "r") as file:
    stock_df = pd.read_json(file, orient="split")
    
print(stock_df.describe())

       daily change
count          6851
unique            6
top          MEDPOS
freq           1753


In [6]:
stock_df.head()

daily change
2000-01-03       BIGNEG
2000-01-04       BIGPOS
2000-01-05       BIGPOS
2000-01-06       BIGPOS
2000-01-07       MEDPOS

In [16]:
def store_time_stats(t, time_bins):
    if t.year not in time_bins:
        time_bins[t.year] = 1
    else: time_bins[t.year] += 1
       
    ym = str(t.year) + "-" + str(t.month)
    if ym not in time_bins:
        time_bins[ym] = 1
    else: time_bins[ym] += 1

In [17]:
#assumes that the stocks are ordered by date

#DEPRECATED
def find_right_timestamp_index(start_ind, time):
    for i in range(start_ind, stock_df.shape[0]):
        date = stock_df.index[i]
        if time.date() != date.date():
            continue
        else: 
            #print ("right date!", date, time)
            return i
    return None

In [18]:
#method to combine the data

def combine_data(df):
    classified = pd.DataFrame(columns=["timestamp", "text", "tweet", "value change"], index=range(df.shape[0]))
    print("classified shape:", classified.shape, "\ntweets_df shape:", df.shape)
    time_bins = {}
    latest_date_ind = 0

    for i, t in enumerate(df["timestamp"]):    
        #count how many tweet per year and month
        store_time_stats(t, time_bins)

        #fetch the change of the tweet's day
        change = stock_df.loc[t.date(), "daily change"]
    
        #save to datastruct
        classified["timestamp"][i] = t
        classified["text"][i] = df["text"][i]
        classified["tweet"][i] = df["tweet"][i]
        classified["value change"][i] = change
    
        if i%500==0: print("progress ", i/classified.shape[0])
            
    return classified, time_bins

In [19]:
classified, time_bins = combine_data(tweets_df)
before_classified, before_time_bins = combine_data(before_tweets_df)
after_classified, after_time_bins = combine_data(after_tweets_df)

classified shape: (13132, 4) 
tweets_df shape: (13132, 3)
progress  0.0
progress  0.03807493146512336
progress  0.07614986293024673
progress  0.11422479439537009
progress  0.15229972586049345
progress  0.19037465732561681
progress  0.22844958879074018
progress  0.26652452025586354
progress  0.3045994517209869
progress  0.34267438318611027
progress  0.38074931465123363
progress  0.418824246116357
progress  0.45689917758148035
progress  0.4949741090466037
progress  0.5330490405117271
progress  0.5711239719768504
progress  0.6091989034419738
progress  0.6472738349070971
progress  0.6853487663722205
progress  0.7234236978373438
progress  0.7614986293024673
progress  0.7995735607675906
progress  0.837648492232714
progress  0.8757234236978373
progress  0.9137983551629607
progress  0.951873286628084
progress  0.9899482180932074
classified shape: (5339, 4) 
tweets_df shape: (5339, 3)
progress  0.0
progress  0.09365049634763065
progress  0.1873009926952613
progress  0.2809514890428919
progress 

In [20]:
import re
#some data about how many tweets per year/month
print ("amount of tweets per year or month\n", time_bins)

s = c = 0
for i, key in enumerate(time_bins):
    if "-" not in str(key):
        continue
    c += 1
    s += time_bins[key]
    
print ("average tweets per month ", s / c)

amount of tweets per year or month
 {2015: 3829, '2015-6': 318, '2015-7': 790, '2015-8': 355, '2015-9': 409, '2015-10': 790, '2015-11': 576, '2015-12': 591, 2016: 4212, '2016-1': 492, '2016-2': 494, '2016-3': 441, '2016-4': 284, '2016-5': 351, '2016-6': 303, '2016-7': 378, '2016-8': 325, '2016-9': 296, '2016-10': 530, '2016-11': 180, '2016-12': 138, 2017: 2637, '2017-1': 214, '2017-2': 154, '2017-3': 147, '2017-4': 152, '2017-5': 155, '2017-6': 212, '2017-7': 245, '2017-8': 267, '2017-9': 320, '2017-10': 325, '2017-11': 250, '2017-12': 196, 2018: 2454, '2018-1': 203, '2018-2': 174, '2018-3': 166, '2018-4': 245, '2018-5': 254, '2018-6': 372, '2018-7': 324, '2018-8': 363, '2018-9': 353}
average tweets per month  328.3


amount of tweets per year or month (before we removed the duplicates from tweets)
 {2015: 3829, '2015-6': 318, '2015-7': 790, '2015-8': 355, '2015-9': 409, '2015-10': 790, '2015-11': 576, '2015-12': 591, 2016: 4460, '2016-1': 492, '2016-2': 494, '2016-3': 441, '2016-4': 284, '2016-5': 351, '2016-6': 303, '2016-7': 378, '2016-8': 325, '2016-9': 296, '2016-10': 530, '2016-11': 290, '2016-12': 276, 2017: 4905, '2017-1': 428, '2017-2': 308, '2017-3': 294, '2017-4': 304, '2017-5': 310, '2017-6': 424, '2017-7': 490, '2017-8': 534, '2017-9': 640, '2017-10': 650, '2017-11': 327, '2017-12': 196, 2018: 2454, '2018-1': 203, '2018-2': 174, '2018-3': 166, '2018-4': 245, '2018-5': 254, '2018-6': 372, '2018-7': 324, '2018-8': 363, '2018-9': 353}
average tweets per month  391.2

In [22]:
before_classified.describe()

timestamp                               text  \
count                  5339                               5339   
unique                 5338                               5173   
top     2016-01-14 05:45:41  makeamericagreatagain trump https   
freq                      2                                 23   

                            tweet value change  
count                        5339         5339  
unique                       5330            6  
top     MAKE AMERICA GREAT AGAIN!       MEDNEG  
freq                            9         1246

In [23]:
# Suffling and splitting the dataframe into train and valid dataframes
from sklearn.model_selection import train_test_split
train, valid = train_test_split(classified, test_size=0.15)
before_train, before_valid = train_test_split(before_classified, test_size=0.15)
after_train, after_valid = train_test_split(after_classified, test_size=0.15)

In [28]:
# Method to make a txt-file out of dataframe for FastText
def save_txt(dataframe, filename):
    learning_data = ""
    for i in range(0, len(dataframe)):
        ind = dataframe.index[i]
        change = dataframe["value change"][ind]
        text = dataframe["text"][ind]
        
        row = "__label__" + change + " " + text
        learning_data += row + "\n"
    print("thousand first characters\n", learning_data[:1000])
    with open(TRAINDATA_DIR + filename, "w") as file:
        file.write(learning_data)

In [29]:
save_txt(train, "train.txt")
save_txt(valid, "valid.txt")
save_txt(before_train, "before_train.txt")
save_txt(before_valid, "before_valid.txt")
save_txt(after_train, "after_train.txt")
save_txt(after_valid, "after_valid.txt")

thousand first characters
 __label__MEDNEG in all fairness to anthony scaramucci he wanted to endorse me st before the republican primaries started but didn t think i was running
__label__MEDNEG just finished the wonderful event on the u s s iowa veterans for a strong america endorsed me such a great honor thank you
__label__BIGPOS thank you hershey pennsylvania get out vote on november th we will maga rallyforriley icymi watch here https
__label__MEDPOS happy birthday https https
__label__BIGNEG joe mcquaid https of the dying union leader wanted ads lunches donations speeches from me and tweets very unethical
__label__SMANEG thank you greeley co real change means restoring honesty to the govt our plan will end govt corruption watch https
__label__MEDPOS i did interview with chris wallace of https in order to be fair he then puts on rove lane and will three trump bashers to discuss
__label__SMAPOS the real story on collusion is in donna b s new book crooked hillary bought the dnc then 

In [30]:
with open(TRAINDATA_DIR + 'valid.json', 'w') as file:
    file.write(valid.to_json(orient='index'))
    
with open(TRAINDATA_DIR + 'before_valid.json', 'w') as file:
    file.write(before_valid.to_json(orient='index'))
    
with open(TRAINDATA_DIR + 'after_valid.json', 'w') as file:
    file.write(after_valid.to_json(orient='index'))

In [31]:
with open(TRAINDATA_DIR + 'valid.json', "r") as file:
    test = pd.read_json(file, orient='index')

In [32]:
test.head()

text           timestamp  \
4012  wsj nbc poll donald trump widens his lead in r... 2016-01-14 23:35:00   
943   thank you https it s time to makeamericagreata... 2015-07-24 20:29:19   
9884  just spoke to governor kenneth mapp of the u s... 2017-09-30 19:30:17   
7852  i cancelled today s meeting with the failing h... 2016-11-22 11:16:45   
4972  rt https i endorse https it s time to unite be... 2016-03-11 15:51:41   

                                                  tweet value change  
4012  WSJ/NBC Poll: Donald Trump Widens His Lead in ...       BIGNEG  
943   Thank you @DennisRodman.  It's time to #MakeAm...       MEDNEG  
9884  Just spoke to Governor Kenneth Mapp of the U.S...       BIGPOS  
7852  I cancelled today's meeting with the failing @...       MEDPOS  
4972  RT @RealBenCarson: I endorse @realDonaldTrump....       SMAPOS

In [33]:
test.describe()

text            timestamp                      tweet value change
count    1970                 1970                       1970         1970
unique   1945                 1967                       1967            6
top     https  2017-09-30 22:26:55  MAKE AMERICA GREAT AGAIN!       MEDNEG
freq        4                    2                          3          453
first     NaN  2015-06-14 21:15:20                        NaN          NaN
last      NaN  2018-09-25 10:53:41                        NaN          NaN